### PR TITLE
feat(TileLayer): add hide skirt parameter

### PR DESCRIPTION
**Pull request by AnthonyGlt**
## Description
Add a new parameter "hideSkirt" to the layer to hide the skirt (borders of tiles).

To test, the parameter can be updated like this ->  view.tileLayer.hideSkirt = true;

## Motivation and Context
Useful when the opacity < 1

## Screenshots (if appropriate)
With hideSkirt = false (current default behavior)
Screenshot from 2024-02-26 11-31-24

with hideSkirt = true
Screenshot from 2024-02-26 11-31-27

# Important
If merged, it's necessary to update the doc to describe this parameter and how to use it

### CODE REVIEW

**Comment by jailln** [comment]
I think we should hide it by default when changing the opacity of the layer to always avoid the effect you show in the first screenshot; WDYT @AnthonyGlt @Desplandis ?

**Comment by AnthonyGlt** [comment]
if we hide it, we see through the terrain , cf screen shot below.

To correct this, it's necessary to switch the scene background color to black and disable the atmosphere

Screenshot from 2024-02-26 14-08-03

**Comment by jailln** [comment]
I meant hide it only when the opacity is < 1

**Comment by AnthonyGlt** [comment]
the result will be nice when we don't zoom but when we zoom, we'll see the background color (like the screenshot above), can be a trade off.

**Comment by jailln** [comment]
> the result will be nice when we don't zoom but when we zoom, we'll see the background color (like the screenshot above), can be a trade off.

Yes you're right, I'm not sure what's best so let's leave it like this for now

**Comment by AnthonyGlt** [comment]
Linked proposal for future development of underground visualization
https://github.com/iTowns/itowns/issues/2281

File: src/Core/Prefab/TileBuilder.js [hunk 1]
```diff
@@ -51,7 +51,9 @@ export default function newTileGeometry(builder, params) {
 
         const geometry = new TileGeometry(params, buffers);
         geometry.OBB = new OBB(geometry.boundingBox.min, geometry.boundingBox.max);
-
+        if (params.hideSkirt) {
+            geometry.setDrawRange(0, geometry.segments * geometry.segments * 2 * 3);//  bufferIndex = (nSeg) * (nSeg) * 2 * 3 (computeBufferTileGeometry.js)
+        }
```
Comment (Desplandis) [inline src/Core/Prefab/TileBuilder.js]: This code is fragile. If there is any future change to the geometry computation in `computeBufferTileGeometry#computeBuffers`, this code would then be invalid.

I have two suggestions:
- Create a function in `computeBufferTileGeometry` to compute the draw range of the tile with and without its skirt. This avoid duplication and could be used at vertex creation and when limiting the draw range.
- Since this a code is specific to the tile geometry,  you could maybe move this code in a `TileGeometry#hideSkirt` setter.

File: src/Core/Prefab/computeBufferTileGeometry.js [hunk 2]
```diff
@@ -22,11 +22,12 @@ export default function computeBuffers(params) {
         //        * u = wgs84.u
         //        * v = textureid + v in builder texture
         uvs: [],
+        segments: params.segments || 8,
```
Comment (Desplandis) [inline src/Core/Prefab/computeBufferTileGeometry.js]: This is not used in the function caller, is this change necessary?

File: src/Layer/TiledGeometryLayer.js [hunk 3]
```diff
@@ -264,6 +275,20 @@ class TiledGeometryLayer extends GeometryLayer {
...
 
+    #hideExistingSkirt(value) {
+        for (const node of this.level0Nodes) {
+            node.traverse((obj) => {
+                if (obj.isTileMesh) {
+                    if (value) {
+                        obj.geometry.setDrawRange(0, obj.geometry.segments * obj.geometry.segments * 2 * 3);  //  bufferIndex = (nSeg) * (nSeg) * 2 * 3 (computeBufferTileGeometry.js)
+                    } else {
+                        obj.geometry.setDrawRange(0, Infinity);
+                    }
```
Comment (Desplandis) [inline src/Layer/TiledGeometryLayer.js]: See previous comment in `TileBuilder.js`, this could be a setter in `TileGeometry`.

File: src/Layer/TiledGeometryLayer.js [hunk 4]
```diff
@@ -71,6 +71,8 @@ class TiledGeometryLayer extends GeometryLayer {
 
         this.protocol = 'tile';
 
+        this.hideSkirt = false;
+
```
Comment (Desplandis) [inline src/Layer/TiledGeometryLayer.js]: By the getter/setter below, this code is not used.

```suggestion
```

File: src/Layer/TiledGeometryLayer.js [hunk 5]
```diff
@@ -98,6 +100,15 @@ class TiledGeometryLayer extends GeometryLayer {
             this.object3d.updateMatrixWorld();
         }));
 
+        let _hideSkirt = this.hideSkirt;
```
Comment (Desplandis) [inline src/Layer/TiledGeometryLayer.js]:
```suggestion
        let _hideSkirt = false;
```

File: src/Layer/TiledGeometryLayer.js [hunk 6]
```diff
@@ -264,6 +275,20 @@ class TiledGeometryLayer extends GeometryLayer {
         return ObjectRemovalHelper.removeChildren(this, node);
     }
 
+    #hideExistingSkirt(value) {
```
Comment (Desplandis) [inline src/Layer/TiledGeometryLayer.js]: Given a recent proposal that I agree with, I think we should use an underscore to mark a function as private. Moreover, since this function is only used once, I think this function could be moved in the `hideSkirt` setter above.

**Review by Desplandis — CHANGES_REQUESTED** [review]
Hi and thanks for your PR ! =)

I approve the proposed addition to the Core API, since the current render is kinda ugly when we need a transparent terrain! We could maybe enforce it (in another PR) when the transparency < 1.0? I dunno.

I have a few architecture comments if you can take a look. =)

File: src/Core/TileGeometry.js [hunk 7]
```diff
@@ -31,6 +31,10 @@ class TileGeometry extends THREE.BufferGeometry {
 
         this.computeBoundingBox();
         this.OBB = {};
+        this.hideSkirt = params.hideSkirt;
```
Comment (AnthonyGlt) [inline src/Core/TileGeometry.js]: Not really satisfied about that. The setter will be called everytime, maybe : 
```suggestion
        if(params.hideSkirt){
           this.hideSkirt = params.hideSkirt;
           }
```
OR

```suggestion
        if(params.hideSkirt){
           this.setDrawRange(0, getBufferIndexSize(this.segments, params.hideSkirt));
           }
```

WDYT @Desplandis  ?

Comment (Desplandis) [reply src/Core/TileGeometry.js]: I agree, we should never have a property and a setter of the same name. I prefer the first solution since we do not duplicate code.

File: src/Layer/TiledGeometryLayer.js [hunk 8]
```diff
@@ -101,6 +102,19 @@ class TiledGeometryLayer extends GeometryLayer {
         this.maxScreenSizeNode = this.sseSubdivisionThreshold * (SIZE_DIAGONAL_TEXTURE * 2);
     }
 
+    get hideSkirt() {
+        return !!this._hideSkirt;
```
Comment (Desplandis) [inline src/Layer/TiledGeometryLayer.js]: Could you please initialize `this._hideSkirt` in the constructor. This eases type checking and avoids the double negation here?

**Review by Desplandis — APPROVED** [review]
Sound good to me, let's merge it! =)
